File: testlua_dl/GameSelectScene.cpp
//
//  GameSelectScene.cpp
//  ddl_plane
//
//  Created by user on 13-10-30.
//
//

#include "GameSelectScene.h"

USING_NS_CC;

GameSelectScene* GameSelectScene::create()
{
    GameSelectScene *pRet = new GameSelectScene();
    if (pRet && pRet->init())
    {
        pRet->autorelease();
        return pRet;
    }
    else
    {
        delete pRet;
        pRet = NULL;
        return NULL;
    }
}

CCScene* GameSelectScene::scene()
{
    // 'scene' is an autorelease object
    CCScene *scene = CCScene::create();
    
    // 'layer' is an autorelease object
    GameSelectScene *layer = GameSelectScene::create();
    
    // add layer as a child to scene
    scene->addChild(layer);
    
    // return the scene
    return scene;
}

// on "init" you need to initialize your instance
bool GameSelectScene::init()
{
    do {
        
        CC_BREAK_IF(!CCLayerGradient::initWithColor(ccc4(0, 0, 0, 120), ccc4(255, 255, 255, 120)));
        
        // Title Label
        CCLabelTTF* pLabel = CCLabelTTF::create(GAME_SELECT_TITLE, GAME_SELET_TITLE_FONT, GAME_SELET_TITLE_SIZE);
        CCSize size = WIN_SIZE;
        pLabel->setPosition( ccp(size.width / 2, size.height - 20) );
        this->addChild(pLabel, 1);
        
        CCSprite* pSprite = CCSprite::create("HelloWorld.png");
        pSprite->setPosition( ccp(size.width/2, size.height/2) );
        
        
        CCMenuItemImage *pCloseItem = CCMenuItemImage::create(
                                                              "CloseNormal.png",
                                                              "CloseSelected.png",
                                                              this,
                                                              menu_selector(GameSelectScene::menuCloseCallback) );
        pCloseItem->setPosition( ccp(CCDirector::sharedDirector()->getWinSize().width - 20, 20) );
        
        CCMenuItemImage *pGame1Item = CCMenuItemImage::create(
                                                              "button_game_start.png",
                                                              "button_press.png",
                                                              this,
                                                              menu_selector(GameSelectScene::menuGame1Enter));
        pGame1Item->setPosition(ccp(size.width / 2, size.height / 2));
        // Game1 item
        CCMenu *pmenu = CCMenu::create(pCloseItem, pGame1Item, NULL);
        pmenu->setPosition(CCPointZero);
        // add the sprite as a child to this layer
        this->addChild(pmenu, 0);
    } while(0);
    return true;
}

void GameSelectScene::menuGame1Enter(CCObject *pSender)
{
    SceneManager::sharedSceneManager()->changeScene(SceneManager::en_GameRunScene);
}

void GameSelectScene::menuCloseCallback(CCObject* pSender)
{
    CCDirector::sharedDirector()->end();
    
#if (CC_TARGET_PLATFORM == CC_PLATFORM_IOS)
    exit(0);
#endif
}

File: testlua_dl/GameSelectScene.lua
--Game select scene
-- require "SceneManager"

function createGameSelectScene()
	local visibleSize = CCDirector:sharedDirector():getVisibleSize()

	local function createScene()
		print "3"
		local function createLayer()
			print "2"
			local function menuCallbackEnterGame()
				SceneManager:changeScene(GAME_RUN_SCENE_TYPE)
			end

			local function createMenu()
				print "1"
				local gameRunItem = CCMenuItemImage:create("button_game_start.png", "button_press.png")
        		gameRunItem:registerScriptTapHandler(menuCallbackEnterGame)
        		gameRunItem:setPosition(visibleSize.width / 2, visibleSize.height / 2)

        		print "1.1"
        		local menu = CCMenu:createWithItem(gameRunItem)
        		print "1.2"
        		print "1.3"
        		menu:setPosition(0, 0);

        		print "create menu"
        		return menu
			end

			local layer = GameSelectScene:create()
			local menu = createMenu()
			layer:addChild(menu)

			print "create layer"
			return layer
		end

		local scene = CCScene:create()
		local layer = createLayer()
		scene:addChild(layer)

		print "create scene"
		return scene
	end

	-- local gameScene = GameSelectScene:scene()
	local gameScene = createScene()
	return gameScene
end